File: runtime/go/runner.go
package engine

import (
	"encoding/json"
	"fmt"
)

// deterministicHandlers holds registered handler functions for deterministic agents.
var deterministicHandlers = map[string]func(map[string]interface{}) (*AgentResult, error){}

var runnerLog = NewLogger("runner")

// RegisterHandler registers a deterministic agent handler by name.
func RegisterHandler(name string, fn func(map[string]interface{}) (*AgentResult, error)) {
	deterministicHandlers[name] = fn
}

// InvokeAgent loads an agent definition and executes it.
func InvokeAgent(agentID string, input interface{}) (*AgentResult, error) {
	def, err := LoadAgent(agentID)
	if err != nil {
		return nil, fmt.Errorf("failed to load agent %s: %w", agentID, err)
	}

	inputMap := ensureMap(input)
	return ExecuteAgent(inputMap, def)
}

// ExecuteAgent runs a single agent (LLM or deterministic) with the given input.
func ExecuteAgent(input map[string]interface{}, agentDef *AgentDefinition) (*AgentResult, error) {
	switch agentDef.Type {
	case "deterministic":
		return executeDeterministic(input, agentDef)
	case "llm":
		return executeLLM(input, agentDef)
	default:
		return nil, fmt.Errorf("unknown agent type: %s", agentDef.Type)
	}
}

func executeDeterministic(input map[string]interface{}, agentDef *AgentDefinition) (*AgentResult, error) {
	handler, ok := deterministicHandlers[agentDef.Handler]
	if !ok {
		return nil, fmt.Errorf("no handler registered for %q", agentDef.Handler)
	}

	runnerLog.Debug("executing deterministic agent", map[string]interface{}{
		"handler": agentDef.Handler,
	})

	return handler(input)
}

func executeLLM(input map[string]interface{}, agentDef *AgentDefinition) (*AgentResult, error) {
	// Build system prompt from the agent's prompt.md content
	systemPrompt := agentDef.Prompt
	if systemPrompt == "" {
		systemPrompt = agentDef.Description
	}

	// Resolve user message template
	userMessage := ResolveTemplate(agentDef.UserMessage, input)

	// Determine schema name
	schemaName := ""
	if agentDef.Schema != nil {
		schemaName = *agentDef.Schema
	}

	// Build user content — may be multipart for image inputs
	var userContent interface{}
	if agentDef.InputType == "image" {
		userContent = buildImageContent(input, userMessage, agentDef.ImageDetail)
	} else {
		userContent = userMessage
	}

	runnerLog.Debug("executing LLM agent", map[string]interface{}{
		"model":      agentDef.Model,
		"schema":     schemaName,
		"input_type": agentDef.InputType,
	})

	return CallLLM(LLMCallOptions{
		Model:        agentDef.Model,
		SystemPrompt: systemPrompt,
		UserContent:  userContent,
		Temperature:  agentDef.Temperature,
		SchemaName:   schemaName,
		Provider:     agentDef.Provider,
	})
}

// buildImageContent creates a multipart content array with an image and text.
func buildImageContent(input map[string]interface{}, textMessage string, detail string) []interface{} {
	if detail == "" {
		detail = "auto"
	}

	parts := []interface{}{}

	// Look for image data in input
	if imageData, ok := input["image"].(string); ok {
		parts = append(parts, map[string]interface{}{
			"type": "image_url",
			"image_url": map[string]interface{}{
				"url":    imageData,
				"detail": detail,
			},
		})
	}
	if imageURL, ok := input["image_url"].(string); ok {
		parts = append(parts, map[string]interface{}{
			"type": "image_url",
			"image_url": map[string]interface{}{
				"url":    imageURL,
				"detail": detail,
			},
		})
	}

	if textMessage != "" {
		parts = append(parts, map[string]interface{}{
			"type": "text",
			"text": textMessage,
		})
	}

	return parts
}

// ensureMap converts input to map[string]interface{}.
// If it is already a map, return it. Otherwise, marshal to JSON and back.
func ensureMap(input interface{}) map[string]interface{} {
	if input == nil {
		return map[string]interface{}{}
	}
	if m, ok := input.(map[string]interface{}); ok {
		return m
	}
	// Try JSON round-trip for struct types
	data, err := json.Marshal(input)
	if err != nil {
		return map[string]interface{}{}
	}
	var m map[string]interface{}
	if err := json.Unmarshal(data, &m); err != nil {
		return map[string]interface{}{}
	}
	return m
}

File: runtime/typescript/src/runner.ts
import type { AgentDefinition, AgentResult, DeterministicHandler } from './types.js';
import { callLLM } from './llm.js';
import { resolveTemplate } from './resolver.js';
import { loadAgent } from './loader.js';
import { createLogger, serializeError } from './logger.js';

const log = createLogger('runner');

// ── Handler registry (user-facing) ──

const handlerRegistry = new Map<string, DeterministicHandler>();

/**
 * Register a deterministic handler function.
 * Called by generated stubs or user code:
 *   registerHandler('product_fetch', (input) => { ... })
 */
export function registerHandler(name: string, fn: DeterministicHandler): void {
	handlerRegistry.set(name, fn);
	log.info(`Handler registered: ${name}`);
}

/** Get all registered handler names (for diagnostics). */
export function getRegisteredHandlers(): string[] {
	return [...handlerRegistry.keys()];
}

// ── Public API (called by generated stubs) ──

/** Invoke a single agent by ID with the given input. */
export async function invokeAgent(
	agentId: string,
	input: Record<string, unknown>
): Promise<AgentResult> {
	log.info(`invokeAgent: ${agentId}`, { input_keys: Object.keys(input) });
	const agentDef = loadAgent(agentId);
	return executeAgent(input, agentDef);
}

// ── Core executor ──

/** Execute an agent (used by orchestrator and invokeAgent). */
export async function executeAgent(
	input: Record<string, unknown>,
	agentDef: AgentDefinition
): Promise<AgentResult> {
	log.info(`Executing agent: ${agentDef.name}`, {
		type: agentDef.type,
		model: agentDef.model,
		input_keys: Object.keys(input),
		input_value_types: Object.fromEntries(
			Object.entries(input).map(([k, v]) => [
				k,
				v === null
					? 'null'
					: v === undefined
						? 'undefined'
						: typeof v === 'string'
							? `string(${v.length})`
							: typeof v,
			])
		),
	});

	if (agentDef.type === 'deterministic') {
		const handlerName = agentDef.handler ?? '';
		const handler = handlerRegistry.get(handlerName);
		if (!handler) {
			log.error(`No deterministic handler registered: ${handlerName}`, {
				available_handlers: [...handlerRegistry.keys()],
			});
			throw new Error(
				`No deterministic handler: ${handlerName}. Available: ${[...handlerRegistry.keys()].join(', ')}`
			);
		}
		try {
			const result = await handler(input);
			log.info(`Deterministic agent complete: ${agentDef.name}`, {
				latency_ms: Math.round(result.metrics.latency_ms),
			});
			return result;
		} catch (err) {
			log.error(`Deterministic agent failed: ${agentDef.name}`, {
				handler: agentDef.handler,
				error: serializeError(err).message,
				input_keys: Object.keys(input),
			});
			throw err;
		}
	}

	return executeLLMAgent(input, agentDef);
}

// ── LLM Agent Executor ──

async function executeLLMAgent(
	input: Record<string, unknown>,
	agentDef: AgentDefinition
): Promise<AgentResult> {
	const systemPrompt = agentDef.prompt ?? '';
	const model = agentDef.model ?? 'gpt-4.1';
	const temperature = agentDef.temperature ?? 0.1;

	if (!systemPrompt) {
		log.warn(`LLM agent ${agentDef.name} has empty system prompt`);
	}

	let userContent:
		| string
		| Array<{ type: string; text?: string; image_url?: { url: string; detail?: string } }>;

	if (agentDef.input_type === 'image') {
		const imageBase64 = input.image as string;
		if (!imageBase64) {
			log.error(`Image agent ${agentDef.name} received no image data`, {
				input_keys: Object.keys(input),
				image_value:
					imageBase64 === undefined
						? 'undefined'
						: imageBase64 === null
							? 'null'
							: 'empty string',
			});
			throw new Error(
				`Image agent "${agentDef.name}" requires input.image but received ${imageBase64 === undefined ? 'undefined' : imageBase64 === null ? 'null' : 'empty string'}`
			);
		}

		log.debug(`Building image message for ${agentDef.name}`, {
			image_length: imageBase64.length,
			detail: agentDef.image_detail,
			user_message: agentDef.user_message?.slice(0, 100),
		});

		userContent = [
			{
				type: 'image_url',
				image_url: {
					url: `data:image/jpeg;base64,${imageBase64}`,
					detail: agentDef.image_detail ?? 'auto',
				},
			},
			{
				type: 'text',
				text: agentDef.user_message ?? 'Analyze this image.',
			},
		];
	} else {
		let message = agentDef.user_message ?? JSON.stringify(input);
		message = resolveTemplate(message, input);

		log.debug(`Built text message for ${agentDef.name}`, {
			message_length: message.length,
			message_preview: message.slice(0, 200),
		});

		userContent = message;
	}

	try {
		const result = await callLLM({
			model,
			systemPrompt,
			userContent,
			temperature,
			schemaName: agentDef.schema,
			base_url: agentDef.base_url,
			api_key_env: agentDef.api_key_env,
		});
		log.info(`LLM agent complete: ${agentDef.name}`, {
			latency_ms: Math.round(result.metrics.latency_ms),
			tokens: { input: result.metrics.input_tokens, output: result.metrics.output_tokens },
		});
		return result;
	} catch (err) {
		log.error(`LLM agent failed: ${agentDef.name}`, {
			model,
			schema: agentDef.schema,
			error: serializeError(err).message,
			error_name: serializeError(err).name,
			error_status: serializeError(err).status,
		});
		throw err;
	}
}
